kv/dist_sender.go

```diff
@@ -531,17 +548,78 @@ func (ds *DistSender) getDescriptors(call proto.Call) (*proto.RangeDescriptor, *
 		// consistency is not required.
 		if call.Args.Header().Txn == nil &&
 			call.Args.Header().ReadConsistency != proto.INCONSISTENT {
-			return nil, nil, &proto.OpRequiresTxnError{}
+			return nil, nil, nil, &proto.OpRequiresTxnError{}
 		}
 		// This next lookup is likely for free since we've read the
 		// previous descriptor and range lookups use cache
 		// prefetching.
 		descNext, err = ds.rangeCache.LookupRangeDescriptor(nextKey, options)
 		if err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
+		}
+	}
+	evict := func() {
+		ds.rangeCache.EvictCachedRangeDescriptor(descKey, desc, isReverse)
+	}
+	return desc, descNext, evict, nil
+}
+
+// Truncate restricts all contained requests to the given key range.
+// Even on error, the returned closure must be executed.
```

Comment:
This comment should explain the closure ("This function modifies *br, and returns a closure that restores it to its original state.")